src/mbgl/text/placement.cpp

```diff
@@ -1268,4 +1154,197 @@ const RetainedQueryData& Placement::getQueryData(uint32_t bucketInstanceId) cons
     return it->second;
 }
 
+class StaticPlacement : public Placement {
+public:
+    explicit StaticPlacement(std::shared_ptr<const UpdateParameters> updateParameters_)
+        : Placement(std::move(updateParameters_), nullopt) {}
+
+private:
+    void commit() override;
+    float symbolFadeChange(TimePoint) const override { return 1.0f; }
+    bool hasTransitions(TimePoint) const override { return false; }
+    bool transitionsEnabled() const override { return false; }
+};
+
+void StaticPlacement::commit() {
+    fadeStartTime = commitTime;
+    for (auto& jointPlacement : placements) {
+        opacities.emplace(
+            jointPlacement.first,
+            JointOpacityState(jointPlacement.second.text, jointPlacement.second.icon, jointPlacement.second.skipFade));
+    }
+}
+class TilePlacement : public StaticPlacement {
+public:
+    explicit TilePlacement(std::shared_ptr<const UpdateParameters> updateParameters_)
+        : StaticPlacement(std::move(updateParameters_)) {}
+
+private:
+    void placeLayers(const RenderLayerReferences&) override;
+    optional<CollisionBoundaries> getAvoidEdges(const SymbolBucket&, const mat4&) override;
+    SymbolInstanceReferences getSortedSymbols(const BucketPlacementData&, float pixelRatio) override;
+    bool stickToFirstVariableAnchor(const CollisionBox& box,
+                                    Point<float> shift,
+                                    const mat4& posMatrix,
+                                    float textPixelRatio) override;
+
+    std::unordered_map<uint32_t, bool> locationCache;
+    optional<CollisionBoundaries> tileBorders;
+    bool onlyLabelsIntersectingTileBorders;
+};
+
+void TilePlacement::placeLayers(const RenderLayerReferences& layers) {
+    // In order to avoid label cut-offs, at first, place the labels,
+    // which cross tile boundaries.
+    onlyLabelsIntersectingTileBorders = true;
```

Comment:
nit: maybe make a wrapper with mandatory enum / bool?

Reply:
this is only used internally, imo a private class member suffices here :thinking: